### PR TITLE
feat: add required field support

### PR DESCRIPTION
**Pull request by attiyaIshaque**
Fields can now be added dynamically to the registration page. These changes are currently behind the `ENABLE_DYNAMIC_REGISTRATION_FIELDS `. 

Ticket: https://openedx.atlassian.net/browse/VAN-838

### CODE REVIEW

File: src/register/RegistrationPage.jsx [hunk 1]
```diff
@@ -543,6 +556,44 @@ class RegistrationPage extends React.Component {
       }
     }
 
+    const formFields = Object.keys(this.props.fieldDescriptions).map((fieldName) => {
+      const fieldData = this.props.fieldDescriptions[fieldName];
+      if (fieldData.name === 'country' && fieldData.type === 'select') {
```
Comment (zainab-amir) [inline src/register/RegistrationPage.jsx:561]: As mentioned by Waheed we can remove the second check
```suggestion
      if (fieldData.name === 'country') {
```

File: src/register/RegistrationPage.jsx [hunk 2]
```diff
@@ -543,6 +556,44 @@ class RegistrationPage extends React.Component {
...
+            handleFocus={this.handleOnFocus}
+            errorMessage={intl.formatMessage(messages['empty.country.field.error'])}
+            handleChange={(value) => this.setState({ country: value })}
+            errorCode={this.state.errorCode}
+            readOnly={this.state.readOnly}
+          />
+        );
+      }
+      if (fieldData.name === 'Honor Code') {
+        return (<HonorCode fieldData={fieldData} value={this.state.values[fieldData.name]} onChangeHandler={this.onChangeHandler} />);
```
Comment (zainab-amir) [inline src/register/RegistrationPage.jsx:580]: nit: If the field is on the same line we don't need to add extra brackets
```suggestion
        return <HonorCode fieldData={fieldData} value={this.state.values[fieldData.name]} onChangeHandler={this.onChangeHandler} />;
```

File: src/register/data/actions.js [hunk 3]
```diff
@@ -53,3 +54,21 @@ export const fetchRealtimeValidationsFailure = () => ({
...
+  type: GET_REQUIRED_FIELDS_DATA.BASE,
+});
+
+export const getRequiredFieldDataBegin = () => ({
+  type: GET_REQUIRED_FIELDS_DATA.BEGIN,
+});
+
+export const getRequiredFieldDataSuccess = (data, extendedProfile) => ({
+  type: GET_REQUIRED_FIELDS_DATA.SUCCESS,
+  payload: { data, extendedProfile }, // remove if not required
```
Comment (zainab-amir) [inline src/register/data/actions.js:69]: is it required?

File: src/register/HonorCode.jsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+          className="opt-checkbox"
+          id={fieldData.name}
+          checked={!!value}
+          name={fieldData.name}
+          value={value}
+          onChange={(e) => onChangeHandler(e)}
+        >
+          <FormattedMessage
+            id="register.page.honor.code"
+            defaultMessage="I agree to the {platformName}  {tosAndHonorCode}"
```
Comment (zainab-amir) [inline src/register/HonorCode.jsx:24]: nit: extra space
```suggestion
            defaultMessage="I agree to the {platformName} {tosAndHonorCode}"
```

**Review by zainab-amir — CHANGES_REQUESTED** [review]
I noticed that we are sending error messages from backend but are not using it here for required fields. We should make sure that appropriate error message is shown if required field is not submitted. 

Also please update `handleSubmit` because currently in code these dynamic fields are not being submitted to backend on submission.

File: src/register/RegistrationPage.jsx [hunk 5]
```diff
@@ -627,27 +754,29 @@ class RegistrationPage extends React.Component {
...
+                        {intl.formatMessage(messages['terms.of.service.and.honor.code'])}
+                      </Hyperlink>
+                    ),
+                    privacyPolicy: (
+                      <Hyperlink variant="muted" destination={getConfig().PRIVACY_POLICY || '#'} target="_blank">
+                        {intl.formatMessage(messages['privacy.policy'])}
+                      </Hyperlink>
+                    ),
+                  }}
+                />
```
Comment (shafqatfarhan) [inline src/register/RegistrationPage.jsx:777]: Can we move this block inside the `HonorCode.jsx` file?

Comment (attiyaIshaque) [reply src/register/RegistrationPage.jsx]: @shafqatfarhan I have changed the code according to your suggestion. please review it.

File: src/field-renderer/FieldRenderer.jsx [hunk 6]
```diff
@@ -6,71 +6,109 @@ import { ExpandMore } from '@edx/paragon/icons';
...
           <Form.Control
             as="select"
             name={fieldData.name}
             value={value}
             onChange={(e) => onChangeHandler(e)}
             trailingElement={<Icon src={ExpandMore} />}
             floatingLabel={fieldData.label}
+            onBlur={handleOnBlur}
+            onFocus={handleFocus}
           >
```
Comment (shafqatfarhan) [inline src/field-renderer/FieldRenderer.jsx:37]: @attiyaIshaque Please add the `aria-invalid` attribute in all the fields that show error state

Ref: https://github.com/openedx/frontend-app-authn/pull/571/files

**Review by shafqatfarhan — COMMENTED** [review]
I have noticed a bunch of UI/UX issues:

Screenshot 2022-05-19 at 1 26 33 PM

- Extra required fields do not have an error (red) border around the field on the active error state

Screenshot 2022-05-19 at 1 38 55 PM

- When the `ENABLE_DYNAMIC_REGISTRATION_FIELDS` flag is false, the honor code paragraph had spacing (margin-top) with the marketing checkbox which is removed now

**Comment by attiyaIshaque** [comment]
@shafqatfarhan I have addressed your comments now
Screenshot 2022-05-19 at 4 02 14 PM

File: src/field-renderer/FieldRenderer.jsx (hunk 6 above)
Comment (attiyaIshaque) [reply src/field-renderer/FieldRenderer.jsx]: Added. :) Thanks for pointing out this.

**Review by shafqatfarhan — APPROVED** [review]
LGTM 👍

**Pull request merged**